### PR TITLE
[WIP] Update workshop llama-3.1 finetuning example and adjust kfto-sft-llm …

### PR DESCRIPTION
**Pull request by abhijeet-dhumal**
## Description

- Update llama-3.1 finetuning workshop readme
- Adjust kfto-sft-llm example with respect to latest kueue-integration updates with step by step instructions

## Merge criteria:

- [ ] The commits are squashed in a cohesive manner and have meaningful messages.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

File: examples/kfto-sft-llm/README.md [hunk 1]
```diff
@@ -17,6 +17,28 @@ It uses HuggingFace SFTTrainer, with PEFT for LoRA and qLoRA, and PyTorch FSDP t
 
 ## Setup
 
+### Configure Kueue (Optional)
+
+> [!NOTE]
+> This section is only required if you plan to use Kueue for workload management (RHOAI 2.21+).
+
+* Update the `nodeLabels` in the `workshops/llm-fine-tuning/kueue/resource_flavor.yaml` file to match your AI worker nodes
```
Comment (astefanutti) [inline examples/kfto-sft-llm/README.md:25]: I wonder if that would be preferable the Kueue configuration be in the example and the workshop points to it or we repurpose the workshop into a first class example, WDYT?

File: examples/kfto-sft-llm/sft.ipynb [hunk 2]
```diff
@@ -300,6 +311,7 @@
     "\n",
     "api_server = \"<API_SERVER>\"\n",
     "token = \"<TOKEN>\"\n",
+    "local_queue = \"<LOCAL_QUEUE_NAME>\" # Optional : replace with your local queue name\n",
```
[the remark this answers is not in the document]
Comment (abhijeet-dhumal) [reply examples/kfto-sft-llm/sft.ipynb:314]: Actually I was planning it for E2E tests where we can use papermill argument for retrieving local-queue name via notebook-CR.. 
Or to keep it as a placeholder for user to provide it similar to `api_server` & `token`

File: examples/kfto-sft-llm/README.md (hunk 1 above)
Comment (abhijeet-dhumal) [reply examples/kfto-sft-llm/README.md]: I will be testing it further thoroughly

File: workshops/kueue/README.md [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+2. Configure authentication:
+   * Get your cluster API server URL and authorization token:
+     * Option 1: Run `oc whoami -t` in your terminal
+     * Option 2: In OpenShift Web console, click your username > "Copy login command"
+
+3. Set up Kueue integration:
+   * In the "Initiate Training Client" cell, enter your team's local-queue name
+   * This enables Kueue workload management for your job
+   > [!NOTE]
+   > For RHOAI versions before 2.21 or to skip Kueue:
```
Comment (Fiona-Waters) [inline workshops/kueue/README.md:188]: Would it be better to remove this part? Since this is a kueue specific workshop? and maybe mention that it is compatible with RHOAI 2.21+

File: examples/kfto-sft-llm/sft.ipynb (hunk 2 above)
[the remark this answers is not in the document]
Comment (abhijeet-dhumal) [reply examples/kfto-sft-llm/sft.ipynb:314]: Yes, If I understood your question correctly you mean when making kueue component managed via RHOAI operator's dsc resource, right?
Yes, the Kueue-VAPB is created and enforces kueue usage out of the box only when Kueue component is managed!

[the remark this answers is not in the document]
Comment (Fiona-Waters) [reply examples/kfto-sft-llm/sft.ipynb:314]: Thanks for confirming, so for this non-kueue example should we have it commented out by default? asking the user to uncomment if they have kueue enabled?

File: examples/kfto-sft-llm/README.md [hunk 4]
```diff
@@ -43,6 +64,19 @@ It uses HuggingFace SFTTrainer, with PEFT for LoRA and qLoRA, and PyTorch FSDP t
 ![](./docs/06.png)
 * Navigate to the `distributed-workloads/examples/kfto-sft-llm` directory and open the `sft` notebook
 
+> [!IMPORTANT]
+> * You will need a Huggingface token with access to the Llama models:
```
Comment (Fiona-Waters) [inline examples/kfto-sft-llm/README.md:68]:
```suggestion
> * You will need a Hugging Face token with access to the Llama models:
```

File: examples/kfto-sft-llm/README.md [hunk 5]
```diff
@@ -43,6 +64,19 @@ It uses HuggingFace SFTTrainer, with PEFT for LoRA and qLoRA, and PyTorch FSDP t
...
+>   * Make sure you have access to the required model (e.g., https://huggingface.co/meta-llama/Llama-3.1-8B)
+>   * Set the `HF_TOKEN` environment variable in your job configuration
+> * If using RHOAI 2.21+, the example supports Kueue integration for workload management:
+>   * When using Kueue:
+>     * Follow the [Configure Kueue](#configure-kueue-optional) section to set up required resources
+>     * Add the local-queue name label to your job configuration to enforce workload management
+>   * You can skip Kueue usage by:
+>     > Note: Kueue Enablement via Validating Admission Policy was introduced in RHOAI-2.21. You can skip this section if using an earlier RHOAI release version.
+>     * Disabling the existing `Kueue-validating-admission-policy-binding`
```
Comment (Fiona-Waters) [inline examples/kfto-sft-llm/README.md:77]:
```suggestion
>     * Disabling the existing `kueue-validating-admission-policy-binding`
```

File: workshops/kueue/README.md [hunk 6]
```diff
@@ -0,0 +1,383 @@
...
+6. Deploy and serve fine-tuned models
+
+## Setup
+
+### Install OpenShift AI
+
+* Log into your OpenShift Web console
+* Go to "Operators" > "OperatorHub" > "AI/Machine Learning"
+* Select the "Red Hat OpenShift AI" operator
+* Install the latest version and Create a default DataScienceCluster resource
```
Comment (Fiona-Waters) [inline workshops/kueue/README.md:39]:
```suggestion
* Install the latest version and create a default DataScienceCluster resource
```

**Review by Fiona-Waters — COMMENTED** [review]
This is excellent! Just a few tiny comments otherwise lgtm. Thanks @abhijeet-dhumal

File: examples/kfto-sft-llm/sft.ipynb [hunk 7]
```diff
@@ -287,6 +287,17 @@
...
    ]
   },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "4e8ac3ef",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# the labels and annotations support for the kubeflow training SDK client's create_job utility was first introduced in v1.9.2\n",
```
[the remark this answers is not in the document]
Comment (abhijeet-dhumal) [reply examples/kfto-sft-llm/sft.ipynb:297]: Actually this note is specifically meant for case when using earlier RHOAI version than 2.21 which has kubeflow training sdk version less than v1.9.2

[the remark this answers is not in the document]
Comment (Fiona-Waters) [reply examples/kfto-sft-llm/sft.ipynb:297]: Ok in that case I think something like this might be clearer, wdyt?
```suggestion
    "# the labels and annotations support for the kubeflow training SDK client's create_job utility was first introduced in v1.9.2, run this cell only if using a RHOAI version earlier than 2.21\n.
```

File: examples/kfto-sft-llm/README.md [hunk 8]
```diff
@@ -43,6 +64,19 @@ It uses HuggingFace SFTTrainer, with PEFT for LoRA and qLoRA, and PyTorch FSDP t
 ![](./docs/06.png)
 * Navigate to the `distributed-workloads/examples/kfto-sft-llm` directory and open the `sft` notebook
 
+> [!IMPORTANT]
+> * You will need a Hugging Face token with access to the Llama models:
```
Comment (ChughShilpa) [inline examples/kfto-sft-llm/README.md:68]: This looks like repetition. This is already mentioned in the notebook itself -
https://github.com/opendatahub-io/distributed-workloads/blob/main/examples/kfto-sft-llm/sft.ipynb

File: examples/kfto-sft-llm/README.md [hunk 9]
```diff
@@ -43,6 +64,19 @@ It uses HuggingFace SFTTrainer, with PEFT for LoRA and qLoRA, and PyTorch FSDP t
...
+>   * Make sure you have access to the required model (e.g., https://huggingface.co/meta-llama/Llama-3.1-8B)
+>   * Set the `HF_TOKEN` environment variable in your job configuration
+> * If using RHOAI 2.21+, the example supports Kueue integration for workload management:
+>   * When using Kueue:
+>     * Follow the [Configure Kueue](#configure-kueue-optional) section to set up required resources
+>     * Add the local-queue name label to your job configuration to enforce workload management
```
Comment (ChughShilpa) [inline examples/kfto-sft-llm/README.md:74]:
```suggestion
>     * Add the 'local-queu-name' to label 'kueue.x-k8s.io/queue-name' to your job configuration to enforce workload management
```

File: workshops/kueue/resources/team2_cluster_queue.yaml [hunk 10]
```diff
@@ -24,5 +24,5 @@ spec:
             - name: memory
               nominalQuota: 2000Gi
             - name: nvidia.com/gpu
-              nominalQuota: '8'
+              nominalQuota: '5'
```
Comment (ChughShilpa) [inline workshops/kueue/resources/team2_cluster_queue.yaml:27]: In setup you mentioned to create this
```
oc apply -f workshops/kueue/resources/team1_cluster_queue.yaml
```

and in notebook, pytorch job needs
```
num_workers=8,
resources_per_worker={
        "nvidia.com/gpu": 1,
        "memory": "64Gi",
        "cpu": 4,
    },
   ```
   So the required total resources doesn't match the resources mentioned in clusterqueue.
   
Did you test this at your end ?

File: examples/kfto-sft-llm/README.md (hunk 9 above)
Comment (abhijeet-dhumal) [reply examples/kfto-sft-llm/README.md:74]: The current wording seems to work well in context and matches what users will see in practice. I afraid the suggested change might create confusion to add given pronoun in single quotes as it is as label value. I'd prefer to keep it as is, but let me know if you feel strongly about it.

**Review by Fiona-Waters — COMMENTED** [review]
/lgtm

File: workshops/kueue/resources/team2_cluster_queue.yaml (hunk 10 above)
Comment (ChughShilpa) [reply workshops/kueue/resources/team2_cluster_queue.yaml:27]: Apologies, i overlooked it

**Review by Fiona-Waters — COMMENTED** [review]
/lgtm

File: workshops/kueue/README.md [hunk 11]
```diff
@@ -0,0 +1,388 @@
...
+oc get pytorchjobs -n <team-namespace>
+
+# Check Kueue workload status
+oc get workloads -n <team-namespace>
+```
+
+```example
+$ oc project team1-project && oc get clusterqueues,resourceflavors,localqueues,pytorchjobs,workloads -o wide
+NAME                                        COHORT         STRATEGY         PENDING WORKLOADS   ADMITTED WORKLOADS
+clusterqueue.kueue.x-k8s.io/cluster-queue                  BestEffortFIFO   0                   0
```
[the remark this answers is not in the document]
Comment (astefanutti) [reply workshops/kueue/README.md]: Just to be clearer, I meant to only remove the default cluster queue which if I understand it correctly should not be there.

[the remark this answers is not in the document]
Comment (abhijeet-dhumal) [reply workshops/kueue/README.md]: Ohh, I see! 👀

**Review by Fiona-Waters — COMMENTED** [review]
/lgtm

File: examples/kfto-sft-llm/README.md [hunk 12]
```diff
@@ -43,8 +48,41 @@ It uses HuggingFace SFTTrainer, with PEFT for LoRA and qLoRA, and PyTorch FSDP t
...
+> * You will need a Hugging Face token if using gated models:
+>   * The examples use gated Llama models that require a token (e.g., https://huggingface.co/meta-llama/Llama-3.1-8B)
+>   * Set the `HF_TOKEN` environment variable in your job configuration
+>   * Note: You can skip the token if switching to non-gated models
+> * If using RHOAI 2.21+, the example supports Kueue integration for workload management:
+>   * When using Kueue:
+>     * Follow the [Configure Kueue (Optional)](#configure-kueue-optional) section to set up required resources
+>     * Add the local-queue name label to your job configuration to enforce workload management
+>   * You can skip Kueue usage by:
+>     > Note: Kueue Enablement via Validating Admission Policy was introduced in RHOAI-2.21. You can skip this section if using an earlier RHOAI release version.
```
Comment (astefanutti) [inline examples/kfto-sft-llm/README.md:61]:
```suggestion
>     > Note: Kueue Enablement via Validating Admission Policy was introduced in RHOAI 2.21. You can skip this section if using an earlier RHOAI release version.
```

File: workshops/kueue/README.md [hunk 13]
```diff
@@ -0,0 +1,539 @@
...
+│                                                              │
+└── ClusterQueue: team2 (ML Team B)                            │
+    ├── Quota: 4x A100 + 6x H100 GPUs, 80 CPU, 1280Gi Memory   │
+    ├── Preemption: Can reclaim from cohort                    │
+    └── Borrowing: Can use unused team1 resources ←────────────┘
+
+Project Layer (Team Access Points)
+├── Team 1 Workloads: (→ team1 ClusterQueue - ML Team A)
+│   ├── LocalQueue: team1-alpha
+│   │   └── Project: team1-project → PyTorchJob: training-workloads
```
Comment (astefanutti) [inline workshops/kueue/README.md:40]: I'd suggest to flip LocalQueue and Project.

Also keep only one LocalQueue per team that's reused by sub-teams.

For example: Team 1 workloads -> Project team1-alpha ->  LocalQueue team1

File: workshops/kueue/resources/resource_flavor.yaml [hunk 14]
```diff
@@ -0,0 +1,23 @@
+apiVersion: kueue.x-k8s.io/v1beta1
```
Comment (astefanutti) [inline workshops/kueue/resources/resource_flavor.yaml:1]: Maybe name the file `resource_flavors.yaml`

File: workshops/kueue/resources/team1_cluster_queue.yaml [hunk 15]
```diff
@@ -17,12 +17,20 @@ spec:
         - memory
         - nvidia.com/gpu
       flavors:
-        - name: nvidia-a100
+        - name: nvidia-a100-80gb
```
Comment (astefanutti) [inline workshops/kueue/resources/team1_cluster_queue.yaml:20]: Maybe move the H100 first, so it's considered first?

File: workshops/kueue/resources/team1_local_queue.yaml [hunk 16]
```diff
@@ -0,0 +1,13 @@
+apiVersion: kueue.x-k8s.io/v1beta1
```
Comment (astefanutti) [inline workshops/kueue/resources/team1_local_queue.yaml:1]: I'd suggest to keep one LocalQueue per team.

File: workshops/kueue/README.md [hunk 17]
```diff
@@ -0,0 +1,539 @@
...
+│ Running   │ Job actively executing on allocated resources                              │
+├───────────┼────────────────────────────────────────────────────────────────────────────┤
+│ Finished  │ Job completed successfully or failed                                       │
+└────────────────────────────────────────────────────────────────────────────────────────┘
+
+
+Resource Flavor Selection Flow:
+==============================
+┌─────────────────────────────┐
+│ PyTorchJob (team1-alpha)    │
```
Comment (astefanutti) [inline workshops/kueue/README.md:127]: It gives the impression team1-alpha is going to get A100 and team1-beta H100 which I don't think is the actual configuration?

File: workshops/kueue/README.md (hunk 13 above)
Comment (abhijeet-dhumal) [reply workshops/kueue/README.md]: Yes that makes perfect sense 🙌 
I have updated the flow accordingly, please review!

**Review by astefanutti — APPROVED** [review]
LGTM!

Awesome work!
